text/0000-template.md

```diff
@@ -0,0 +1,341 @@
+
+This flood strategy is important to guarantee that all peers will receive the new blocks/transactions as soon as possible. New blocks received in this phase cannot have a parents between the genesis and the last checkpoint.
+
+## Stateful Validation
+
+As we may happen to download blocks and transactions that does not reach the genesis, we cannot guarantee we will have all parents and inputs to fully validate the blocks and transactions. So, we split the validation in the following states:
+
+1. `not-validated`: initial state, no validation has been made yet.
+
+2. `basic-validated`: all graph fields (except the parents and inputs) are valid, including the PoW, the weight field, the timestamp, etc. Funds fields have not been verified yet.
+
+3. `full-validated`: the block/transaction has been fully validated and reaches the genesis.
+
+4. `invalid`: this block/transaction is invalid and might be discarded (could be kept for quick hash detection or for banning the original peer that propagated it).
+
+Notice that `full-validated` requires the block/transaction to reach the genesis to get to this validation state. It is a critical part of this state because it guarantees that all blocks and transactions behind it are `full-validated` as well.
+
+The valid transitions between the states are shown in the figure below:
+
+![fig-state-diagram](./img/fig-state-diagram.svg "State Diagram")
+
+A block/transaction can go from `not-validated` to `invalid` for different reasons (e.g. invalid PoW, invalid timestamp, no outputs, no inputs). These blocks/transactions might be immediately discarded.
+
+The common reasons that moves a block/transaction from `basic-validated` to `invalid` are (i) the funds are unbalanced and fails, and (ii) any of their parents or inputs are invalid. The invalid block/transactions can be discarded any time. **XXX Should we do it during the node initialization?**
+
+All transactions of the mempool must be `full-validated`. If they can't be fully validated, they can be safely discarded.
+
+We will rarely have a `not-validated` block/transaction because they are basically verified as soon as they are received. No transaction/block saved in the storage might be `not-validated`.
+
+## Optimizations & Sources of blocks and transactions
+
+The fullnode receives new blocks and transactions from two different sources: (i) the Sync Protocol, and (ii) its APIs.
+
+When the source is the Phase 1 of the Sync Protocol, we may optimize and only verify the PoW. This optimization assumes that nobody can generate another transaction with the same hash as the original transaction (because it would imply finding a sha256 conflict). The fullnode currently does not use this optimization and verifies every block/transaction received.
+
+When a transaction is received by the API, it must be fully validated. If it is not possible, the fullnode must reject it.
+
+## Measure of Sync Progress
```

Comment:
I would also add here what's the definition for "is_synced" in this new algorithm. Maybe add this in another section but I think it's an important idea.

Reply:
I guess it is here: "We say that we are synced to a peer when all blocks between the genesis and the best block have been fully downloaded. The mempool does not be affect the sync status.".

I also added a complement for being virtually synced. Can you review it again, please?

Reply:
My bad, I think I might've missed this. But there is a typo, I think. 

> The mempool does not ~~be~~ affect the sync status

Reply:
Fixed. Thanks!